### ISSUE TITLE
twistd --uid 'myuid' fails even though the uid is the one of the current user

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3172]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3172|
|Type|defect|
|Created|2008-04-08 14:19:16Z|
|Branch|https://github.com/twisted/twisted/tree/switchuid-current-3172-2|

The traceback:
```
Traceback (most recent call last):
  File "./bin/twistd", line 21, in <module>
    run()
  File "/home/therve/Projets/Twisted/trunk/twisted/scripts/twistd.py", line 27, in run
    app.run(runApp, ServerOptions)
  File "/home/therve/Projets/Twisted/trunk/twisted/application/app.py", line 614, in run
    runApp(config)
  File "/home/therve/Projets/Twisted/trunk/twisted/scripts/twistd.py", line 23, in runApp
    _SomeApplicationRunner(config).run()
  File "/home/therve/Projets/Twisted/trunk/twisted/application/app.py", line 337, in run
    self.postApplication()
  File "/home/therve/Projets/Twisted/trunk/twisted/scripts/_twistd_unix.py", line 207, in postApplication
    self.startApplication(self.application)
  File "/home/therve/Projets/Twisted/trunk/twisted/scripts/_twistd_unix.py", line 309, in startApplication
    self.shedPrivileges(self.config['euid'], uid, gid)
  File "/home/therve/Projets/Twisted/trunk/twisted/scripts/_twistd_unix.py", line 281, in shedPrivileges
    switchUID(uid, gid, euid)
  File "/home/therve/Projets/Twisted/trunk/twisted/python/util.py", line 664, in switchUID
    initgroups(uid, gid)
  File "/home/therve/Projets/Twisted/trunk/twisted/python/util.py", line 641, in initgroups
    _setgroups_until_success(l)
  File "/home/therve/Projets/Twisted/trunk/twisted/python/util.py", line 587, in _setgroups_until_success
    setgroups(l)
OSError: [Errno 1] Operation not permitted
```

The cause: initgroups use grp.grpall to check all the groups of the user. However, as noted in the grp documentation:
```
Note that most users are not explicitly listed as members of the group
they are in according to the password database. Check both databases
to get complete membership information.
```

So the group returned by `pwd.getpwuid` needs to be added to the list of available groups.

<details><summary>Searchable metadata</summary>

```
trac-id__3172 3172
type__defect defect
reporter__therve therve
priority__low low
milestone__ 
branch__branches_switchuid_current_3172_2 branches/switchuid-current-3172-2
branch_author__therve__mesozoic therve, mesozoic
status__closed closed
resolution__fixed fixed
component__core core
keywords__twistd twistd
time__1207664356000000 1207664356000000
changetime__1297049350000000 1297049350000000
version__None None
owner__mesozoic mesozoic
cc__exarkun cc__glyph cc__mesozoic
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [23309]) Branching to 'initgroups-self-uid-3172'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [23310]) Fix the problem by adding the group

Refs [#3172](https://github.com/twisted/twisted/issues/3172)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

I think the branch is good, please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

Maybe there could be an assertion at the end of that test?  Like about what the current GID is and/or what the secondary groups are?

Also, please update copyright date and give test_util a module docstring.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [26976]) Branching to 'initgroups-self-uid-3172-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

I've added an assertion, it might be better than nothing.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

  1. `pwd.getpwuid` returns an object with nice attributes.  If you want, replace the tuple unpacking in `util.initgroups` with attribute access.
  1. It seems that `initgroups(uid, gid)` succeeds (or rather, fails silently) when called with a gid other than the current gid (when the current uid is not privileged).  eg, in the branch:

```
$ python -c 'from twisted.python.util import initgroups
> initgroups(1000, 2000)
> '
$
```

vs in trunk:

```
$ python -c 'from twisted.python.util import initgroups
initgroups(1000, 2000)
'
Traceback (most recent call last):
  File "<string>", line 2, in <module>
  File "twisted/python/util.py", line 641, in initgroups
    _setgroups_until_success(l)
  File "twisted/python/util.py", line 587, in _setgroups_until_success
    setgroups(l)
OSError: [Errno 1] Operation not permitted
$
```

In general, the various things `initgroups` now does confuse me a bit.  It could do with some more comments explaining each step it takes, I think.

Is `initgroups` the best place to fix this misbehavior of twistd?  I'm sure there's some subtlety involved that I'm missing, but it seems like not trying to change uid/gid at _all_ if the uid specified on the command line is the current uid would also fix this, and more simply?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [26983]) Branching to 'switchuid-current-3172'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

OK, I've tried another approach by fixing switchUID this time. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @therve|
|-|-|


Thanks so much for fixing this.  It generally looks pretty good.

  1. Given that `switchUID` is changing, this would be an excellent opportunity to add a docstring, don't you think? :)  Specifically, one that explains the behavior in the you're-not-root case (given the behavior implemented in the branch) would be good.
    1. From what I can tell, `initgroups` won't actually work without root permissions anyway, but it does create a change in semantics between the root and non-root versions of setting your UID.  My understanding is that if switchUID is called in a process as root, it will pick up new groups that a user has been added to since the process started, but if it's called in the case where it's a no-op, it will leave the process with whatever groups it had when it started up.  I don't believe it's possible to pick up those new groups without root permissions (`/usr/bin/newgrp` on my system is u+s, I believe for this reason) but I feel it's important to document this subtle distinction.
  1. Things that patch `os` (or `sys`, or `warnings`, or `types`, or other similarly broadly shared modules) directly in tests are dangerous.  I'd like to discourage them.  In ascending order of preference:
    1. Patch just `os` on the SUT as it is imported. In this case, `self.patch(util, "os", self.mockos)`.
    1. However, patching something as broad as `twisted.python.util` (which is used by trial) is fraught with its own perils.  If this were a class, I'd say make it a class or instance attribute; you could still put an attribute on the function, of course.
    1. You could do something tricky by calling `types.FunctionType` with a globals object of your own design, and the code object from `switchUID`.  This would allow you to create a locally patched version for the test and call that, rather than mutating the global version.
    1. The best option, of course, is just to pass parameters.  This is, I'm sure, what exarkun would suggest :).  Personally, I'm a bit ambivalent about suggesting this option in a case like this, because it pollutes the implementation code with testing concerns (the signature would end up looking like `switchUID(uid, gid, euid=False, os=os, initgroups=initgroups)`, which would be a difficult signature to support as a public interface.  (Why not `(uid, gid, euid=False, os=os, pwd=pwd)`, since that is what initgroups is built out of?  What about future support for `spwd` or some other, fancier user database API, etc, etc?)  However, this has the advantage of being very simple to follow and not relying on any test-harness trickery which would have to be implemented and tested itself, so I think it's still the best option, just not ideal.

I am confident that I'll be happy with whatever resolution you think is most appropriate to these points, so if you're comfortable landing this without another review after dealing with this feedback, go ahead.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

I'm worried that the semantic change here will result in confusion.  Why don't we just report an error nicely on the command line if a non-root user asks twistd to change uid or gid?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

OK, probably.

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_14"></a>|@mesozoic commented|
|-|-|

(In [28702]) Branching to 'switchuid-current-3172-2'

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_15"></a>|@mesozoic commented|
|-|-|

I believe this is ready for review.

 * Added a docstring to switchUID 
 * Added a call to warnings.warn in switchUID 
 * Limited the scope of patching the os module in test_util.SwitchUIDTest 
 * Added tests to test_util.SwitchUIDTest for warnings and fixed behavior 
 * Make UnixApplicationRunner.shedPrivileges fail gracefully when switchUID actually does raise an OSError

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @mesozoic|
|-|-|

Thanks for picking this up.

For clarity, the behavior implemented in the branch when passing `--uid <current uid>`

```
2010-04-11 11:54:43-0400 [-] Log opened.
2010-04-11 11:54:43-0400 [-] twistd 10.0.0+r28819 (/usr/bin/python 2.6.4) starting up.
2010-04-11 11:54:43-0400 [-] reactor class: twisted.internet.selectreactor.SelectReactor.
2010-04-11 11:54:43-0400 [-] twisted.web.server.Site starting on 8080
2010-04-11 11:54:43-0400 [-] Starting factory <twisted.web.server.Site instance at 0x9b2a06c>
2010-04-11 11:54:43-0400 [-] /home/exarkun/Projects/Twisted/trunk/twisted/python/util.py:711: exceptions.UserWarning: ignoring call to switchUID(uid=1000, gid=None, euid=0) because uid is already 1000
2010-04-11 11:54:43-0400 [-] set uid/gid 1000/None
^C2010-04-11 11:54:47-0400 [-] Received SIGINT, shutting down.
2010-04-11 11:54:47-0400 [-] (Port 8080 Closed)
2010-04-11 11:54:47-0400 [-] Stopping factory <twisted.web.server.Site instance at 0x9b2a06c>
2010-04-11 11:54:47-0400 [-] Main loop terminated.
2010-04-11 11:54:47-0400 [-] Server Shut Down.
```

And for `--uid <another uid>` when the current uid is not root:

```
2010-04-11 11:56:37-0400 [-] Log opened.
2010-04-11 11:56:37-0400 [-] twistd 10.0.0+r28819 (/usr/bin/python 2.6.4) starting up.
2010-04-11 11:56:37-0400 [-] reactor class: twisted.internet.selectreactor.SelectReactor.
2010-04-11 11:56:37-0400 [-] twisted.web.server.Site starting on 8080
2010-04-11 11:56:37-0400 [-] Starting factory <twisted.web.server.Site instance at 0x912f08c>
2010-04-11 11:56:37-0400 [-] failed to set uid/gid 1001/None (are you root?) -- exiting.
```

I think I'd be happier to just see the latter behavior in all cases.  However, I recall that this issue was discussed at a recent sprint and some decisions were made there, but I don't recall what the decisions were.  If there was some motivation for the currently implemented behavior, it'd be great to document it (even if only on this ticket).  The feature is subtle enough that forcing everyone to independently re-discover why it is correct is probably unreasonable.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_17"></a>|@mesozoic commented|
|-|-|

I can't speak to the motivations behind originally opening the ticket; it's not one I created and it was opened a while ago. I was simply trying to help clean up the ticket queue and saw "low hanging fruit".

It seems intuitive that ```twistd --uid <current-uid>``` ought to work even if the user isn't root. I can't think of a use case where the correct behavior would be a failure. But either way, the current behavior is to fail with a traceback, which I think is less helpful than the error message stating "failed to set uid/gid 1001/None (are you root?)"

Does that address your question? If the core point of contention is that "twistd --uid <current-uid> actually should fail", it sounds like this ticket is simply invalid and ought to be shelved so that folks like myself don't pick it up again.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

Replying to [mesozoic](#note_17):
> I can't speak to the motivations behind originally opening the ticket; it's not one I created and it was opened a while ago. I was simply trying to help clean up the ticket queue and saw "low hanging fruit".

Upon reading his comment in detail, I don't think exarkun was really asking for the _original_ motivation, just for _some_ motivation.  Specifically: there should be documentation _included as part of the change_, since we are changing the behavior and the behavior should be clearly documented and explained.  Everyone's personal opinions about the feature aside, whatever the feature finally implements really should be documented.

> It seems intuitive that ```twistd --uid <current-uid>``` ought to work even if the user isn't root. I can't think of a use case where the correct behavior would be a failure. But either way, the current behavior is to fail with a traceback, which I think is less helpful than the error message stating "failed to set uid/gid 1001/None (are you root?)"

I agree.

As a general rule, in Twisted, and _especially_ in _twistd_, I think we should try to soldier on as much as possible, with the notable exception of cases where it's a potential security problem to keep running.  This is admittedly an edge case, but the security properties requested are the same as those being granted, except for the part where the server has _fewer_ privileges upon startup (rather than being root -> you, it's just you when it starts up).

My motivation for saying this is that it should be possible for applications to implement 'self-healing' configurations, so if they screw up something like this, they can notice and fix it so that it's correct for the next restart.  But an application will have a hard time fixing its own configuration if twistd gives up and dies before actually invoking the application's code.

As a minor nit though, I think I might prefer it if the message were a bit more generally phrased.  I don't really care about calls to `switchUID` (especially since I can see that it's `switchUID` by looking at the file/line combo present in the warning itself); it should say something more like "tried to drop privileges and set UID to x, but UID was already x.  continuing, but perhaps you should start as root next time?"  OK, that phrasing is a little awkward, but hopefully you get the idea.

> Does that address your question? If the core point of contention is that "twistd --uid <current-uid> actually should fail", it sounds like this ticket is simply invalid and ought to be shelved so that folks like myself don't pick it up again.

I don't think anybody's saying that the _current_ behavior (a traceback) is desirable.  I think that the behavior in the branch is pretty good, but I think exarkun is suggesting 'a warning, then exit' rather than 'a warning, then continue'.  As I said, I disagree with that view, but perhaps if he explained _why_ he'd be happier to see that behavior in all cases I would change my mind...

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_20"></a>|@mesozoic set owner to @glyph|
|-|-|

Reassigning to glyph (because he told me to) now that I've committed recommended changes to the warning message in [30477]. Please hash out intended functionality with exarkun.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_22"></a>|@jerub set owner to @mesozoic|
|-|-|

Fantastic, this is a lovely patch and has all of the above review comments addressed adequately. The one blocker is a discussion about the correct and/or intended behaviour here, there's been no discussion on the ticket about this, and I agree with glyph's opinion presented in this ticket that the behaviour in this branch is pretty good.

I like the warning text, it seems quite clear.

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_23"></a>|@mesozoic set status to `closed`|
|-|-|

(In [30634]) Merge switchuid-current-3172-2: twistd will no longer fail if a non-root user passes --uid 'myuid' as a command-line argument. Instead, it will emit an error message.

Authors: therve, mesozoic
Reviewers: jerub, glyph
Fixes: [#3172](https://github.com/twisted/twisted/issues/3172)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set status to `reopened`|
|-|-|

(In [30635]) <<EOF
Revert r30634 - test suite regression

Reopens [#3172](https://github.com/twisted/twisted/issues/3172)

This test failure appears on most platforms:

```
===============================================================================
[ERROR]
Traceback (most recent call last):
  File /var/lib/buildbot/twisted/lucid32-py2.7maint/Twisted/twisted/python/test/test_util.py, line 209, in test_currentEUID
    self.assertIn('but uid is already %i' % uid, warnings[0]['message'])
exceptions.NameError: global name 'uid' is not defined

twisted.python.test.test_util.SwitchUIDTest.test_currentEUID
-------------------------------------------------------------------------------
```
EOF

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_25"></a>|@mesozoic commented|
|-|-|

Sorry, should have checked all test suites (not just twisted.test) -- not sure how I missed that earlier. Running buildbots before re-merging.

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_26"></a>|@mesozoic set status to `closed`|
|-|-|

(In [30638]) Merge switchuid-current-3172-2: twistd will no longer fail if a non-root user passes --uid 'myuid' as a command-line argument. Instead, it will emit an error message.

Authors: therve, mesozoic 

Reviewers: jerub, glyph 

Fixes: [#3172](https://github.com/twisted/twisted/issues/3172)
